src/server/session.ts

```diff
@@ -2912,41 +2934,55 @@ namespace ts.server {
             let request: protocol.Request | undefined;
             let relevantFile: protocol.FileRequestArgs | undefined;
             try {
-                request = <protocol.Request>JSON.parse(message);
-                relevantFile = request.arguments && (request as protocol.FileRequest).arguments.file ? (request as protocol.FileRequest).arguments : undefined;
+                Debug.assert(!tracing.canPop());

```

Comment:
Same comments as above about the nested `try-catch`, and the while loop etc.

Reply:
Same answer - I believe cancellation can occur without throwing (but also without popping the event stack).

Reply:
I put the asserts back and they didn't fire, so I've dropped the try-finally.